pkg/northbound/proto/roles.proto

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2019-present Open Networking Foundation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+package proto;
+
+// DeviceRoleConfig carries the template configuration associated with a device role
+message DeviceRoleConfig {
+    string role = 1;
+    DeviceProperties properties = 2;
+    DevicePipeline pipeline = 3;
+}
+
+// DeviceProperties is a set of initial configuration properties to be applied to a device.
+message DeviceProperties {
+    repeated DeviceProperty props = 1;
+}
+
+// DeviceProperty is a path/type/value tuple
+message DeviceProperty {
+    string path = 1;
+    string type = 2;
+    string value = 3;
+}
+
+// DevicePipeline carries information about the P4 pipeline configuration
+message DevicePipeline {
+    string pipeline = 1;
+}
+
+// DeviceRoleChange is an event describing a change to a device role configuration.
+message DeviceRoleChange {
+    enum ChangeType {
+        UPDATED = 0;
+        ADDED = 1;
+        DELETED = 2;
+    }
+    ChangeType change = 1;
+    DeviceRoleConfig config = 2;
+}
+
+// DeviceRoleChangeRequest is a request for a change to a device role configuration
+message DeviceRoleChangeRequest {
+    enum ChangeType {
+        UPDATE = 0;
+        ADD = 1;
+        DELETE = 2;
+    }
+    DeviceRoleConfig config = 2;
+}
+
+// DeviceRoleChangeResponse is a response for a change to a device role configuration
+message DeviceRoleChangeResponse {
+    int32 status = 1;
+    DeviceRoleChange change = 2;
+}
+
+// DeviceRoleRequest is a request for device role configuration.
+message DeviceRoleRequest {
+    string role = 1;
+}
+
+// DeviceRoleService provides means for setting up device role configurations
+// in support of zero-touch provisioning activities.
+service DeviceRoleService {
```

Comment:
the similarity of this to gNMI (with the CRUD operations) make me wonder if we should be defining this as a YANG model instead and treating it like a config - not of a device but of onos-config itself - a self management model

Reply:
This has crossed my mind as well, but given the fact that we may need to incorporate not just gNMI-based configs, but also P4Runtime based ones and possible whatever the RAN subsystem may add on beyond these two, I thought we ought to for now keep it separate and its own thing.  I purposely chose this form of API to vet-out Carmelo's suggestion of using this style of API for the topology subsystem.

Reply:
sounds logical